common/cache/lru.go

```diff
@@ -319,6 +314,21 @@ func (c *lru) deleteInternal(element *list.Element) {
 	delete(c.byKey, entry.key)
 }
 
+func (c *lru) evictOnceInternal() {
+	element := c.byAccess.Back()
+	for element != nil {
+		entry := element.Value.(*entryImpl)
+		if entry.refCount == 0 {
+			c.deleteInternal(element)
+			return
+		}
+
+		// entry.refCount > 0
```

Comment:
I noticed there's no guard in `Release` to prevent decrementing the counter to less than zero: https://github.com/temporalio/temporal/pull/2353/files#diff-2d61d82d9a8acd1031ca31021b40eb5f8766840ac54b4113dba6b50e8e033afeR241

albeit unlikely - a bug could cause the refCount to be less than zero. In which case this assertion would not be correct (refCount > 0).

Reply:
the ref count must be maintained correctly, by the caller (meaning if double release ever happen, the behavior is undefined)

the reference count > 0 means whether the cache item is being used, e.g. cache cannot evict this item if ref count > 0

the reason to use a refcount here is to guarantee all mutation to a mutable state must be serialized
* caller fist get the mutable state cache entry
* caller then use the mutable state (from the cache entry) by first locking it

about double release, there are few things that can be done
* if ref count < 0, panic (this should be long term solution, not suitable for patch)
* do <= 0 check instead of == 0 check, this will prevent some weird behavior, but when refcount < 0, system is already broken.

Reply:
synced offline, this PR is ok for patching

Reply:
Agreed! Looks good to me.